### PR TITLE
mgr/dashboard: monitoring:Implement BlueStore onode hit/miss counters into the dashboard

### PR DESCRIPTION
**Pull request by aaSharma14**
Provide the details pulled from Bluestore stats in order to display the onode hit/miss counters

Depends on #44334
Fixes: https://tracker.ceph.com/issues/53577
Signed-off-by: Aashish Sharma <aasharma@redhat.com>

Screenshot from 2022-01-05 13-20-05
Screenshot from 2022-01-05 13-20-05(1)

<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by avanthakkar** [comment]
Some unit-test would be nice to have @aaSharma14

**Comment by aaSharma14** [comment]
> Some unit-test would be nice to have @aaSharma14

Done. Thanks @avanthakkar

**Comment by pereman2** [comment]
jenkins test make check

**Review by pereman2 — APPROVED** [review]
Looks good

**Comment by aaSharma14** [comment]
jenkins test make check

**Comment by epuertat** [comment]
@neha-ojha @ljflores @markhpc folks, what do you think about this one?

**Comment by alfonsomthd** [comment]
jenkins test dashboard

**Comment by aaSharma14** [comment]
jenkins test dashboard

**Review by ljflores — APPROVED** [review]
Looks great!

I want to mention that there is another PR open right now (#44334) that shortens the names of the BlueStore onode hit/miss counters when printed. I don't think that would affect this PR, but linking it just in case.

**Comment by aaSharma14** [comment]
> Looks great!
> 
> I want to mention that there is another PR open right now (#44334) that shortens the names of the BlueStore onode hit/miss counters when printed. I don't think that would affect this PR, but linking it just in case.

Thanks @ljflores , Yes, in this case we need to change the metrics name here in this PR as well. We can merge this one after #44334 gets merged.

**Review by avanthakkar — APPROVED** [review]

**Comment by ljflores** [comment]
@aaSharma14 the branch needs to be updated

File: monitoring/grafana/dashboards/jsonnet/grafana_dashboards.jsonnet
Comment (epuertat) [inline monitoring/grafana/dashboards/jsonnet/grafana_dashboards.jsonnet:770]: No need to escape double quotes. That's inherited from the JSON code, but jsonnet supports both double and single quotes, which is the one you're using here:
```suggestion
        addTargetSchema('ceph_bluestore_bluestore_onode_hits{ceph_daemon=~"$osd"}', 1, 'time_series', '$osd hits')
```

File: monitoring/grafana/dashboards/tests/features/osd-device-details.feature
Comment (epuertat) [inline monitoring/grafana/dashboards/tests/features/osd-device-details.feature:6]: `ceph_bluestore_bluestore_onode_hits` ok, I think I now understand that the clean-up in the RADOS perf-counters was to remove this redundancy, @ljflores ? :S

File: monitoring/grafana/dashboards/jsonnet/grafana_dashboards.jsonnet
Comment (epuertat) [inline monitoring/grafana/dashboards/jsonnet/grafana_dashboards.jsonnet:779]: When introducing jsonnet/grafonnet, one of the key deciding factors was increasing maintainability/readability of the Grafana dashboard JSON code, so why not making things more readable? It's hard to find where a PromQL expression ends and the next starts (btw nested `label_replace()` should be a code smell if such a thing may exist in PromQL 🙈 ):

```suggestion
        'Physical Device Latency for $osd',
        '',
        's',
        'Read (-) / Write (+)',
        '(label_replace(irate(node_disk_read_time_seconds_total[1m]) / irate(node_disk_reads_completed_total[1m]),
  "instance", "$1", "instance", "([^:.]*).*") and on (instance, device) label_replace(  label_replace(ceph_disk_occupation{ceph_daemon=~"$osd"}, "device", "$1", "device", "/dev/(.*)"), "instance", "$1", "instance", "([^:.]*).*"))',
        '(label_replace(irate(node_disk_write_time_seconds_total[1m]) / irate(node_disk_writes_completed_total[1m]), "instance", "$1", "instance", "([^:.]*).*") and on (instance, device) label_replace(label_replace(ceph_disk_occupation{ceph_daemon=~"$osd"}, "device", "$1", "device", "/dev/(.*)"), "instance", "$1", "instance", "([^:.]*).*"))',
        '{{instance}}/{{device}} Reads', '{{instance}}/{{device}} Writes', 0, 21, 6, 9)
```

**Review by epuertat — COMMENTED** [review]
@aaSharma14 can you please do a clean-up on escaped double quotes `\"` so we are progressively improving our grafonnet code base?

File: monitoring/grafana/dashboards/tests/features/osd-device-details.feature
Comment (ljflores) [reply monitoring/grafana/dashboards/tests/features/osd-device-details.feature:6]: @epuertat yes, and I believe #44334 mainly affects the output of `ceph tell osd.* perf dump` by changing the name of the perf counter and without modifying the perf counter "key".

File: monitoring/grafana/dashboards/jsonnet/grafana_dashboards.jsonnet
Comment (aaSharma14) [reply monitoring/grafana/dashboards/jsonnet/grafana_dashboards.jsonnet]: Done. Thanks @epuertat

File: monitoring/grafana/dashboards/jsonnet/grafana_dashboards.jsonnet
Comment (aaSharma14) [reply monitoring/grafana/dashboards/jsonnet/grafana_dashboards.jsonnet]: Done. Thanks @epuertat

File: monitoring/grafana/dashboards/tests/features/osd-device-details.feature
Comment (epuertat) [reply monitoring/grafana/dashboards/tests/features/osd-device-details.feature:6]: @ljflores I'm afraid that the name (label( of the perf counter is actually the key (I don't see any other string literal over there, and AFAIK C++ doesn't support reflection... yet). Anyway, as I commented in that PR, given this 2 are the only metrics left with this redundant format, let's go for it.

**Review by epuertat — APPROVED** [review]
LGTM @aaSharma14 ! But you need to update the metric names based on #44334.

**Comment by aaSharma14** [comment]
jenkins test dashboard cephadm

**Comment by ljflores** [comment]
@epuertat sounds good

**Review by neha-ojha — APPROVED** [review]
looks great!

**Comment by epuertat** [comment]
Now I check the time-series chart again, being both series monotonic counters, I realize that if the hit/miss ratio is uneven (let's say 90% hit ratio, or 90% miss ratio), over time the smallest metric will be almost zero, and a temporal chart won't be very informative.



A possibility would be to display the hit-miss ratio over time. There's this [Grafana Stat Panel widget](https://grafana.com/docs/grafana/v7.0/panels/visualizations/stat-panel/) that we don't use, and allows us printing a percentage (e.g.: hit ratio), a trend in the background, and also a threshold-driven color.

image

And coming back to the pie chart, we could replace it with the [built-in gauge panel](https://grafana.com/docs/grafana/v7.0/panels/visualizations/gauge-panel/) (which is great for displaying a binary dataset). One issue with pie-charts (vs. doughnut charts or gauges), is that for conveying information they rely on areas, and areas are squared values, so they actually display information^2. Gauges also allow defining thresholds, so we can visually tell users whether a 20% hit rate is actually a good (green) or bad (red) value.

image


So, summarizing my point:
* Can we have 1 new chart instead of 2?
* Can that be a gauge (if we don't mind temporal trends) or a Stat Panel otherwise?

**Comment by aaSharma14** [comment]
> Now I check the time-series chart again, being both series monotonic counters, I realize that if the hit/miss ratio is uneven (let's say 90% hit ratio, or 90% miss ratio), over time the smallest metric will be almost zero, and a temporal chart won't be very informative.
> 
> 
> 
> A possibility would be to display the hit-miss ratio over time. There's this [Grafana Stat Panel widget](https://grafana.com/docs/grafana/v7.0/panels/visualizations/stat-panel/) that we don't use, and allows us printing a percentage (e.g.: hit ratio), a trend in the background, and also a threshold-driven color.
> 
> image
> 
> And coming back to the pie chart, we could replace it with the [built-in gauge panel](https://grafana.com/docs/grafana/v7.0/panels/visualizations/gauge-panel/) (which is great for displaying a binary dataset). One issue with pie-charts (vs. doughnut charts or gauges), is that for conveying information they rely on areas, and areas are squared values, so they actually display information^2. Gauges also allow defining thresholds, so we can visually tell users whether a 20% hit rate is actually a good (green) or bad (red) value.
> 
> image
> 
> So, summarizing my point:
> 
> * Can we have 1 new chart instead of 2?
> * Can that be a gauge (if we don't mind temporal trends) or a Stat Panel otherwise?

Thank you for the suggestion @epuertat . I am currently implementing a gauge panel for this one. In any case what is the ideal onode Hits ratio ? so we can display a warning or error scenario if the ratio goes below that value. Eg. I have kept it 50% as of now, so if the ratio goes below 50%, it would show that in red color and otherwise green. @ljflores @neha-ojha Do you guys have some insight about this?

Screenshot from 2022-01-04 15-31-39

**Comment by ljflores** [comment]
> In any case what is the ideal onode Hits ratio ?

@aaSharma14 @epuertat Just talked with @aclamk about this. He said he uses the threshold of 75%, where below that is fine, and above that is something to investigate.

This is looking great!

File: monitoring/grafana/dashboards/osd-device-details.json
Comment (aaSharma14) [reply monitoring/grafana/dashboards/osd-device-details.json]: Thanks @neha-ojha , I have modified the panels. PTAL again

**Comment by aaSharma14** [comment]
> > In any case what is the ideal onode Hits ratio ?
> 
> @aaSharma14 @epuertat Just talked with @aclamk about this. He said he uses the threshold of 75%, where above that is fine, and below that is something to investigate.
> 
> This is looking great!
> 
> Edit: I switched the "above" and "below" scenarios. Apologies, the comment is correct now.

Thankyou @ljflores for the information, I have updated the PR accordingly.

**Comment by aaSharma14** [comment]
jenkins test dashboard

**Comment by aaSharma14** [comment]
jenkins test make check

**Comment by aaSharma14** [comment]
jenkins test api

**Comment by aaSharma14** [comment]
jenkins test make check

**Review by alfonsomthd — APPROVED** [review]

**Pull request merged**
